viewer/packages/cad-geometry-loaders/src/sector/SectorDownloadScheduler.ts

```diff
@@ -0,0 +1,107 @@
+/*!
+ * Copyright 2022 Cognite AS
+ */
+
+import { ConsumedSector, LevelOfDetail, WantedSector } from '@reveal/cad-parsers';
+import log from '@reveal/logger';
+import { DeferredPromise } from '@reveal/utilities/src/DeferredPromise';
+import assert from 'assert';
+
+export type SectorDownloadData = {
+  sector: WantedSector;
+  downloadSector: (sector: WantedSector) => Promise<ConsumedSector>;
+};
+
+export class SectorDownloadScheduler {
+  private readonly _maxConcurrentSectorDownloads: number;
+
+  private readonly _pendingSectorDownloads: Map<string, Promise<ConsumedSector>>;
+  private readonly _queuedSectorDownloads: Map<
+    string,
+    [WantedSector, (sector: WantedSector) => Promise<ConsumedSector>, DeferredPromise<ConsumedSector>]
+  >;
+  private readonly _sectorDownloadQueue: string[];
+
+  constructor(maxConcurrentSectorDownloads = 20) {
+    this._maxConcurrentSectorDownloads = maxConcurrentSectorDownloads;
+    this._pendingSectorDownloads = new Map();
+    this._queuedSectorDownloads = new Map();
+    this._sectorDownloadQueue = [];
+  }
+
+  public queueSectorBatchForDownload(downloadData: SectorDownloadData[]): Promise<ConsumedSector>[] {
+    return downloadData.map(sectorDownloadData => {
+      const { sector, downloadSector } = sectorDownloadData;
+      const sectorIdentifier = `${sector.metadata.id}-${sector.modelIdentifier}`;
+      const pendingSector = this._pendingSectorDownloads.get(sectorIdentifier);
+
+      if (pendingSector !== undefined) {
+        return pendingSector;
+      }
+
+      if (this._pendingSectorDownloads.size < this._maxConcurrentSectorDownloads) {
+        return this.addSectorToPendingDownloads(downloadSector, sector, sectorIdentifier);
+      }
+
+      const queuedSector = this._queuedSectorDownloads.get(sectorIdentifier);
+
+      if (queuedSector !== undefined) {
+        const [_0, _1, downloadPromise] = queuedSector;
+        return downloadPromise;
+      }
+
+      const deferredSectorDownloadPromise = new DeferredPromise<ConsumedSector>();
+      this._sectorDownloadQueue.push(sectorIdentifier);
+      this._queuedSectorDownloads.set(sectorIdentifier, [sector, downloadSector, deferredSectorDownloadPromise]);
+
+      return deferredSectorDownloadPromise;
+    });
```

Comment:
This could maybe be a separate function to make it clear that either it adds the request for immediate download, or it creates a deferred one